src/extra/gfx/texture.cpp

```diff
@@ -0,0 +1,169 @@
+#include "../gfx.hpp"
+#include <gfx/texture.hpp>
+#include <gfx/error_utils.hpp>
+#include <params.hpp>
+#include <stdexcept>
+
+using namespace shards;
+namespace gfx {
+
+enum ComponentType {
+  Float,
+  Int8,
+  Int16,
+};
+
+struct TextureFormatException : public std::runtime_error {
+  TextureFormatException(ComponentType componentType, Types::TextureType_ asType) : std::runtime_error(formatError(componentType, asType)) {}
+
+  static std::string formatError(ComponentType componentType, Types::TextureType_ asType) {
+    return fmt::format("Image with component type '{}' can not be converted to texture type '{}'",
+                      magic_enum::enum_name(componentType), magic_enum::enum_name(asType));
+  }
+};
+
+struct TextureShard {
+  static inline shards::Types InputTypes{{CoreInfo::ImageType}};
+  static SHTypesInfo inputTypes() { return InputTypes; }
+  static SHTypesInfo outputTypes() { return Types::Texture; }
+  static SHOptionalString help() { return SHCCSTR("Creates a texture from an image"); }
+
+  TexturePtr texture;
+  OwnedVar textureVar;
+
+  PARAM_VAR(asType_, "Type", "Type to interpret image data as. (Default: UNormSRGB for RGBA8 images, UNorm for other formats)",
+            {Types::TextureType, CoreInfo::NoneType});
+  PARAM_IMPL(TextureShard, PARAM_IMPL_FOR(asType_));
+
+  void warmup(SHContext *context) { PARAM_WARMUP(context); }
+  void cleanup() { PARAM_CLEANUP(); }
+
+  SHVar activate(SHContext *shContext, const SHVar &input) {
+    auto &image = input.payload.imageValue;
+
+    textureVar.valueType = SHType::Object;
+    textureVar.payload.objectTypeId = Types::TextureTypeId;
+    textureVar.payload.objectVendorId = gfx::VendorId;
+    textureVar.payload.objectValue = &texture;
+
+    if (!texture)
+      texture = std::make_shared<Texture>();
```

Comment:
I would move those 2 lines above for better readability. It feels a bit weird to pass a pointer to `&texture` before it is actually initialized (is it actually correct anyway?).

Reply:
It's a pointer to the shared_ptr (which is owned by the shard) so it's correct but I can move it above